src/devices/bus/spectrum/zxbus.h

```diff
@@ -0,0 +1,124 @@
+    *22      WR │19│   │19│ BUSRQ
+     -not-used- │20│   │20│ RES     *26
+    *24    WAIT │21│   │21│ A7      *37
+     -not-used- │22│   │22│ A6      *36
+     -not-used- │23│   │23│ A5      *35
+    *27      M1 │24│   │24│ A4
+          RFRSH │25│   │25│ CSR  (CSROM)
+    *38      A8 │26│   │26│ BUSAK
+    *40     A10 │27│   │27│ A9      *39
+  KAY=+5V  n.u. │28│   │28│ A11     *1
+  KAY=+12V n.u. │29│   │29│ n.u. KAY=+5V
+            GND │30│   │30│ GND
+                │  │   │  │
+                │ B│   │A │
+                └──┴───┴──┘
+  '-' - not used
+  '*' - Z80 CPU out
+
+*******************************************************************************************/
+
+#ifndef MAME_BUS_SPECTRUM_ZXBUS_H
+#define MAME_BUS_SPECTRUM_ZXBUS_H
+
+#pragma once
+
+class zxbus_slot_device : public device_t, public device_slot_interface
+{
+public:
+	zxbus_slot_device(machine_config const &mconfig, char const *tag, device_t *owner, uint32_t clock);
+
+	template <typename T, typename U>
+	zxbus_slot_device(machine_config const &mconfig, char const *tag, device_t *owner, u32 clock, T &&zxbus_tag, U &&slot_options, const char *default_option, bool fixed)
+		: zxbus_slot_device(mconfig, tag, owner, clock)
+	{
+		option_reset();
+		slot_options(*this);
+		set_default_option(default_option);
+		set_fixed(fixed);
+        m_zxbus_bus.set_tag(std::forward<T>(zxbus_tag));
+	}
+
+protected:
+    zxbus_slot_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, uint32_t clock);
+
+	virtual void device_start() override;
+
+	required_device<device_t> m_zxbus_bus;
+};
+
+DECLARE_DEVICE_TYPE(ZXBUS_SLOT, zxbus_slot_device)
+
+
+class zxbus_device : public device_t
+{
+public:
+    zxbus_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock);
+
+	template <typename T> void set_iospace(T &&tag, int spacenum) { m_iospace.set_tag(std::forward<T>(tag), spacenum); }
+	template<typename T> void install_device(offs_t addrstart, offs_t addrend, T &device, void (T::*map)(class address_map &map), uint64_t unitmask = ~u64(0))
+	{
+		m_iospace->install_device(addrstart, addrend, device, map, unitmask);
+	}
+
+	void add_slot(const char *tag);
+	void add_slot(device_slot_interface *slot);
+
+protected:
+	zxbus_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, uint32_t clock);
+
+	virtual void device_start() override {};
+	virtual void device_reset() override {};
+
+    required_address_space m_iospace;
+    std::forward_list<device_slot_interface *> m_slot_list;
+};
+
+DECLARE_DEVICE_TYPE(ZXBUS, zxbus_device)
+
+
+class device_zxbus_card_interface : public device_interface
+{
+public:
+	void set_zxbus_device();
+    void set_zxbusbus(device_t *zxbus_device) { m_zxbus_dev = zxbus_device; }
+
+protected:
+    device_zxbus_card_interface(const machine_config &mconfig, device_t &device);
+
+	zxbus_device *m_zxbus;
+	device_t     *m_zxbus_dev;
```

Comment:
Why not have a single `void set_zxbusbus(zxbus_device &bus_dev) { m_zxbus = &bus; }` and get rid of the extra member and the downcasts that come with it?